OpenRA.Mods.Common/Traits/Air/Aircraft.cs

```diff
@@ -521,20 +530,44 @@ public void ModifyDeathActorInit(Actor self, TypeDictionary init)
 
 		void INotifyBecomingIdle.OnBecomingIdle(Actor self)
 		{
-			if (Info.VTOL && Info.LandWhenIdle)
+			OnBecomingIdle(self);
+		}
+
+		protected virtual void OnBecomingIdle(Actor self)
+		{
+			var atLandAltitude = self.World.Map.DistanceAboveTerrain(CenterPosition) == Info.LandAltitude;
+
+			// Work-around to prevent players from accidentally canceling resupply by pressing 'Stop',
+			// by re-queueing ResupplyAircraft as long as resupply hasn't finished and aircraft is still on resupplier.
+			// TODO: Investigate moving this back to ResolveOrder's "Stop" handling,
+			// once conflicts with other traits' "Stop" orders have been fixed.
+			if (atLandAltitude)
+			{
+				var host = GetActorBelow();
+				if (host != null && (CanRearmAt(host) || CanRepairAt(host)))
+				{
+					self.QueueActivity(new ResupplyAircraft(self));
+					return;
+				}
+			}
+
+			if (!atLandAltitude && Info.VTOL && Info.LandWhenIdle)
 			{
 				if (Info.TurnToLand)
 					self.QueueActivity(new Turn(self, Info.InitialFacing));
 
 				self.QueueActivity(new HeliLand(self, true));
 			}
-			else if (!Info.CanHover)
+			else if (!Info.CanHover && !atLandAltitude)
 				self.QueueActivity(new FlyCircle(self, -1, Info.IdleTurnSpeed > -1 ? Info.IdleTurnSpeed : TurnSpeed));
-
-			// Temporary HACK for the AutoCarryall special case (needs CanHover, but also HeliFlyCircle on idle).
-			// Will go away soon (in a separate PR) with the arrival of ActionsWhenIdle.
+			else if (atLandAltitude && (Info.TakeOffOnResupply || ReservedActor == null))
+				self.QueueActivity(new TakeOff(self));
 			else if (Info.CanHover && self.Info.HasTraitInfo<AutoCarryallInfo>() && Info.IdleTurnSpeed > -1)
+			{
+				// Temporary HACK for the AutoCarryall special case (needs CanHover, but also HeliFlyCircle on idle).
+				// Will go away soon (in a separate PR) with the arrival of ActionsWhenIdle.
 				self.QueueActivity(new HeliFlyCircle(self, Info.IdleTurnSpeed > -1 ? Info.IdleTurnSpeed : TurnSpeed));
+			}
```

Comment:
I'm really looking forward to #15873, this whole chain of conditionals is so horrible :rofl:

Reply:
Yeah. I'm actually thinking about pulling in #15803 first because less activities + properly untangled CanHover/VTOL would make #15873 a bit simpler than it would currently be.